File: backend/app/services/geocode.py
# app/services/geocode.py

import aiohttp  # For making async HTTP requests
import logging
from app.config import OPENCAGE_API_KEY  # Your OpenCage API key stored securely

# Set up logging to help track issues during development or debugging
logger = logging.getLogger(__name__)

# Asynchronously fetch latitude and longitude for a given address
async def geocode_address(address: str):
    try:
        # Create an async session to call the OpenCage API
        async with aiohttp.ClientSession() as session:
            url = "https://api.opencagedata.com/geocode/v1/json"
            params = {
                "q": address,         # The address to search for
                "key": OPENCAGE_API_KEY  # Your API key
            }
            
            # Send GET request to the API
            async with session.get(url, params=params) as resp:
                # If API request fails, return an error
                if resp.status != 200:
                    return None, f"Geocoding failed: HTTP {resp.status}"

                # Parse the JSON response
                result = await resp.json()

                # If we received valid results, extract the coordinates
                if result["results"]:
                    geometry = result["results"][0]["geometry"]
                    return (geometry["lat"], geometry["lng"]), None  # Success: return coordinates
                return None, "Invalid address"  # Address not found

    except Exception as e:
        # Catch any unexpected errors during the process
        return None, f"Geocoding error: {str(e)}"

File: frontend/src/services/geocode.js
// src/services/geocode.js

/**
 * getCoords
 * This function takes an address string and returns geographic coordinates using
 * the free OpenStreetMap Nominatim API.
 *
 * @param {string} address - The human-readable address to geocode
 * @returns {Promise<{lat: number, lon: number} | null>} The coordinates or null if not found
 */
export async function getCoords(address) {
  // Build the API request URL with the address encoded for safe HTTP transmission
  const url = `https://nominatim.openstreetmap.org/search?format=json&q=${encodeURIComponent(address)}`;

  // Send the request to OpenStreetMap's Nominatim API
  const res = await fetch(url);

  // Parse the JSON response
  const data = await res.json();

  // If no results are found, return null
  if (data.length === 0) return null;

  // Extract latitude and longitude from the first result
  return {
    lat: parseFloat(data[0].lat),
    lon: parseFloat(data[0].lon),
  };
}
